main/sources/scripture-app-builder.ts

```diff
@@ -1,18 +1,286 @@
+    sfmJson: SfmJson,
+    xmlDoc: Document,
+    projectName: string,
+    directory: string
+  ): BKChapter[] {
+    const chapters = [];
+
+    // there are two different schemas
+    let isPageSchema = true;
+
+    let chapterNumbers = _.map(
+      xmlDoc.querySelectorAll("book[id='" + bookId + "'] > page[num] > audio > timing-filename"),
+      (n) => n.parentElement?.parentElement?.attributes.getNamedItem('num')?.value ?? undefined
+    ).filter((cn) => cn != null && cn !== '');
+    if (chapterNumbers.length === 0) {
+      // attempt to use old schema
+      isPageSchema = false;
+      chapterNumbers = _.map(
+        xmlDoc.querySelectorAll("book[id='" + bookId + "'] > audio[chapter] > timing-filename"),
+        (n) => n.parentElement?.attributes.getNamedItem('chapter')?.value ?? undefined
+      ).filter((cn) => cn != null && cn !== '');
+    }
+    if (chapterNumbers.length === 0) {
+      return [];
+    }
+
+    for (const chapterNumber of chapterNumbers) {
+      const fileSelector = isPageSchema
+        ? "book[id='" + bookId + "'] > page[num='" + chapterNumber + "'] > audio > filename"
+        : "book[id='" + bookId + "'] > audio[chapter='" + chapterNumber + "'] > filename";
+
+      const timingFileSelector = isPageSchema
+        ? "book[id='" + bookId + "'] > page[num='" + chapterNumber + "'] > audio > timing-filename"
+        : "book[id='" + bookId + "'] > audio[chapter='" + chapterNumber + "'] > timing-filename";
+
+      const audioFileName = xmlDoc.querySelector(fileSelector)?.textContent;
+      const audioFileLength = xmlDoc.querySelector(fileSelector)?.getAttribute('len');
+      const timingFileName = xmlDoc.querySelector(timingFileSelector)?.textContent;
+      // cancel if anything can't be found
+      if (!audioFileName || !audioFileLength || !timingFileName || !chapterNumber) {
+        return [];
+      }
+
+      const audioFile = this.calculateAudioFile(audioFileName, directory, projectName);
+      if (!audioFile) {
+        return [];
+      }
+
+      const segments = this.makeSegments(sfmJson, +chapterNumber, timingFileName, projectName, directory);
+      if (segments.length > 0) {
+        chapters.push({
+          name: chapterNumber,
+          audio: { files: [{ filename: audioFile, length: parseInt(audioFileLength) }] },
+          segments: segments,
+        });
+      }
+    }
+
+    return chapters;
+  }
+
+  private makeSegments(
+    sfmJson: SfmJson,
+    chapterNumber: number,
+    timingFileName: string,
+    projectName: string,
+    directory: string
+  ): BKSegment[] {
+    const segments = [];
+
+    const timingFileFullName = path.join(directory, projectName, projectName + '_data', 'timings', timingFileName);
+
+    let timingFileContent;
+
+    try {
+      timingFileContent = fs.readFileSync(timingFileFullName, 'utf8');
+    } catch (error) {
+      console.error(error);
+      return [];
+    }
+
+    const timingPattern = /(\d+\.?\d*)\t(\d+\.?\d*)\t(\d+)/gs;
```

Comment:
RegEx are tricky things. It is preferable to add a comment stating what the intended purpose/result is.

Reply:
added some clarification

Reply:
Nice